core/src/main/java/org/pentaho/platform/workitem/WorkItemLifecycleEvent.java

```diff
@@ -0,0 +1,270 @@
+import org.apache.commons.lang.builder.ToStringBuilder;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.pentaho.platform.util.ActionUtil;
+
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.util.Date;
+import java.util.Map;
+import java.util.UUID;
+
+/**
+ * This class encapsulates all information pertaining to a "work item" at a specific point in its lifecycle.
+ */
+public class WorkItemLifecycleEvent {
+
+  private static final Log logger = LogFactory.getLog( WorkItemLifecycleEvent.class );
+
+  private String workItemUid;
+  private String workItemDetails;
+  private WorkItemLifecyclePhase workItemLifecyclePhase;
+  private String lifecycleDetails;
+  private Date sourceTimestamp;
+  private String sourceHostName;
+  private String sourceHostIp;
+  private static String HOST_NAME;
+  private static String HOST_IP;
+
+  static {
+    try {
+      HOST_NAME = InetAddress.getLocalHost().getCanonicalHostName();
+      HOST_IP = InetAddress.getLocalHost().getHostAddress();
+    } catch ( final UnknownHostException uhe ) {
+      logger.error( uhe.getLocalizedMessage() );
+    }
+  }
+
+  /**
+   * Default constructor, needed for serialization purposes.
+   */
+  protected WorkItemLifecycleEvent() {
+  }
+
+  /**
+   * Creates the {@link WorkItemLifecycleEvent} with all the required parameters. The {@code sourceTimestmap} is
+   * set to the current date.
+   *
+   * @param workItemUid     a {@link String} containing unique identifier for the {@link WorkItemLifecycleEvent}
+   * @param workItemDetails a {@link String} containing details of the {@link WorkItemLifecycleEvent}
+   */
+  public WorkItemLifecycleEvent( final String workItemUid, final String workItemDetails ) {
+    this( workItemUid, workItemDetails, new Date() );
+  }
+
+  /**
+   * Creates the {@link WorkItemLifecycleEvent} with all the required parameters.
+   *
+   * @param workItemUid     a {@link String} containing unique identifier for the {@link WorkItemLifecycleEvent}
+   * @param workItemDetails a {@link String} containing details of the {@link WorkItemLifecycleEvent}
+   * @param sourceTimestamp a {@link Date} representing the time the lifecycle change occured.
+   */
+
+  public WorkItemLifecycleEvent( final String workItemUid, final String workItemDetails, final Date sourceTimestamp ) {
+    this.workItemUid = workItemUid;
+    this.workItemDetails = workItemDetails;
+    this.sourceTimestamp = sourceTimestamp;
+
+    // if the workItemUid is null, generate it
+    if ( this.workItemUid == null ) {
+      this.workItemUid = generateWorkItemId();
+    }
+    // Set sourceTimestamp to current date only if not already provided
+    if ( this.sourceTimestamp == null ) {
+      this.sourceTimestamp = new Date();
+    }
+
+    // set the default values for host name and ip, they can be changed directly if needed
+    this.sourceHostName = HOST_NAME;
```

Comment:
Yes, this is where the WI ran. We have not ran full tests to see what the values are on w WN, but will do so become merging. We can certainly report on container specific variables, I will look into it, but we will not have access to those variables on the server side, where other lifecycle events related to the same work item might be reported, so I'm not sure it makes sense to include those vars here... We can discuss this offline.